text/0000-named-arguments.md

```diff
@@ -0,0 +1,618 @@
+
+## API compatibility
+
+Making a positional argument named is a backwards compatible change. However, changing or removing the argument name is a breaking change. Only the argument _pattern_ can be changed backwards compatibly.
+
+For example, the following changes are backwards compatible:
+
+```rust
+fn foo(a: i32) {}
+fn foo(.a: i32) {}
+fn foo(.a b: i32) {}
+fn foo(.a _: i32) {}
+```
+
+Argument names can be added to positional arguments _from right to left_, without breaking backwards compatibility:
+
+```rust
+fn foo(a: i32, b: i32) {}
+fn foo(a: i32, .b: i32) {}
+fn foo(.a: i32, .b: i32) {}
+```
+
+Argument names can also be added to trait functions backwards compatibly:
+
+```rust
+trait Foo {
+    fn foo(a: i32, b: i32);
+}
+// -->
+trait Foo {
+    fn foo(.a: i32, .b: i32);
+}
+```
+
+## ABI compatibility
+
+Named arguments don't affect the ABI or code generation in any way. For example, you can use named arguments in `extern "C"` functions.
+
+## Interactions with the type system
+[interactions-with-the-type-system]: #interactions-with-the-type-system
+
+Named arguments don't interact with the type system, since named arguments are _not_ part of the function's type. For example, the following is invalid:
+
+```rust
+fn foo(.a: i32, .b: i32) {}
+
+let f: fn(.a: i32, .b: i32) = foo; // forbidden!
+let f: fn(i32, i32)         = foo; // correct
+
+fn higher_order(f: impl Fn(.a: i32, .b: i32)) {} // forbidden!
+fn higher_order(f: impl Fn(i32, i32)) {}         // correct
+```
+
+Also, named arguments are _not_ involved in type inference.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Minor: Verbosity
+
+In function calls, the syntax is slightly more verbose than in other languages such as Python or Swift:
+
+```rust
+// Swift
+greet(person: "Tim", alreadyGreeted: true)
+// Rust
+greet(.person = "Tim", .already_greeted = true)
+```
+
+## Minor: Named arguments aren't mandatory
+
+Because named arguments aren't mandatory in function calls, this might encourage bad habits where people don't specify named arguments out of laziness. This is especially bad when the function names are short and unspecific, as in the example above.
+
+However, this can also be seen as a good thing: If named arguments were mandatory, API authors might be hesitant to add named arguments, since they make function calls more verbose. However, if you can choose at every function call if you want to specify argument names, there are no downsides for an API author in making function arguments named.
+
+The main reason why named arguments aren't mandatory is that existing APIs can be converted to use named arguments backwards compatibly. This would not be possible if named arguments were mandatory.
+
+### Remedy: Add a clippy lint
+
+Laziness to specify named arguments can be remedied by adding a clippy lint that warns when a named argument is omitted.
+
```

Comment:
A planned lint checking for full or none used argument names would be more specific. You describe the intended behavior of the lint also below.

Reply:
I'm not sure what you have in mind, can you explain?

Reply:
In your latest revision you say what the clippy lint should do in the clippy lint section. So this is okay now.